support/arr/arr.go

```diff
@@ -0,0 +1,633 @@
+	return nil, ErrNoImplementation
+}
+
+// ExplodePluckParameters Explode the "value" and "key" arguments passed to "pluck".
+func ExplodePluckParameters[T any](array []T, key ...int) ([]T, error) {
+	return nil, ErrNoImplementation
+}
+
+// Map Run a map over each of the items in the array.
+func Map[T, U any](arr []T, fn func(T, int) U) []U {
+	res := make([]U, len(arr))
+	for i, v := range arr {
+		res[i] = fn(v, i)
+	}
+	return res
+}
+
+// Prepend the given value to the beginning of an array or associative array.
+func Prepend[T any](arr []T, value T) []T {
+	return append([]T{value}, arr...)
+}
+
+// Pull Get a value from the array, and remove it.
+func Pull[T any](arr []T, key int, def T) ([]T, T, error) {
+	v, _ := Get(arr, key, def)
+
+	res, err := Forget(arr, key)
+	return res, v, err
+}
+
+// Query Convert the array into a query string.
+func Query[T any](m []T) (*string, error) {
+	return nil, ErrNoImplementation
+}
+
+// Random returns one or a specified number of random values from a slice.
+func Random[T any](arr []T, number *int) ([]T, error) {
+	//rand.Seed(time.Now().UnixNano())
+
+	requested := 1
+	if number != nil {
+		requested = *number
+	}
+
+	count := len(arr)
+
+	if requested > count {
+		return nil, ErrExceedMaxLength
+	}
+
+	if number == nil {
+		return []T{arr[rand.Intn(count)]}, nil
+	}
+
+	if requested == 0 {
+		return []T{}, nil
+	}
+
+	indices := rand.Perm(count)[:requested]
+
+	res := make([]T, 0, requested)
+	for _, index := range indices {
+		res = append(res, arr[index])
+	}
+
+	return res, nil
+}
+
+// Set an array item to a given value using int key
+// todo: thread safe?
+func Set[T any](arr *[]T, key int, value T) error {
+	if key < 0 {
+		return ErrInvalidKey
+	}
+
+	if key >= len(*arr) {
+		newSlice := make([]T, key+1)
+		copy(newSlice, *arr)
+		*arr = newSlice
+	}
+
+	(*arr)[key] = value
+	return nil
+}
+
+// Shuffle the given array and return the result.
+func Shuffle[T any](arr []T, seed *int64) []T {
+	res := make([]T, len(arr))
+	copy(res, arr)
+
+	var r *rand.Rand
+	if seed == nil {
+		randSeed := time.Now().UnixNano()
+		r = rand.New(rand.NewSource(randSeed))
+	} else {
+		r = rand.New(rand.NewSource(*seed))
+	}
+	r.Shuffle(len(res), func(i, j int) { res[i], res[j] = res[j], res[i] })
+	return res
+}
+
+// Sort the nested array using the given callback.
+// todo: generic
+func Sort(arr []any, fn func(i, j int) bool) []any {
+	if len(arr) == 0 {
+		return arr
+	}
+
+	sort.Slice(arr, func(i, j int) bool {
+		return fn(i, j)
+	})
+
+	for i, v := range arr {
+		switch val := v.(type) {
+		case []any:
+			arr[i] = Sort(val, fn)
+		default:
+		}
+	}
+
+	return arr
+}
+
+// Sort the nested array in descending order using the given callback.
+// todo: generic
+func SortDesc(arr []any, fn func(i, j int) bool) []any {
+	return Sort(arr, fn)
+}
+
+// SortRecursive Recursively sort an array by values.
+// todo: generic
+func SortRecursive(arr []any, descending bool) ([]any, error) {
```

Comment:
S: A great method, but if we can't use generic or will more complex, I think we can not implement it first.